code/espurna/telnet.ino

```diff
@@ -120,13 +127,90 @@ void _telnetData(unsigned char clientId, void *data, size_t len) {
     #if TERMINAL_SUPPORT
         terminalInject(data, len);
     #endif
+}
+
+void _telnetNotifyConnected(unsigned char i) {
+
+    DEBUG_MSG_P(PSTR("[TELNET] Client #%u connected\n"), i);
+
+    // If there is no terminal support automatically dump info and crash data
+    #if TERMINAL_SUPPORT == 0
+        info();
+        wifiDebug();
+        crashDump();
+        crashClear();
+    #endif
+
+    #ifdef ESPURNA_CORE
+        _telnetClientsAuth[i] = true;
+    #else
+        _telnetClientsAuth[i] = !_telnetAuth;
+        if (_telnetAuth) {
+            if (getAdminPass().length()) {
+                _telnetWrite(i, "Password: ");
+            } else {
+                _telnetClientsAuth[i] = true;
+            }
+        }
+    #endif
+
+    _telnetFirst = true;
+    wifiReconnectCheck();
 
 }
 
-void _telnetNewClient(AsyncClient *client) {
+#if TELNET_SERVER == TELNET_SERVER_WIFISERVER
+void _telnetLoop() {
+    if (_telnetServer.hasClient()) {
+        int i;
 
-    if (client->localIP() != WiFi.softAPIP()) {
+        for (i = 0; i < TELNET_MAX_CLIENTS; i++) {
+            if (!_telnetClients[i] || !_telnetClients[i]->connected()) {
+
+                _telnetClients[i] = std::unique_ptr<WiFiClient>(new WiFiClient(_telnetServer.available()));
+
+                if (_telnetClients[i]->localIP() != WiFi.softAPIP()) {
+                    // Telnet is always available for the ESPurna Core image
+                    #ifdef ESPURNA_CORE
+                        bool telnetSTA = true;
+                    #else
+                        bool telnetSTA = getSetting("telnetSTA", TELNET_STA).toInt() == 1;
+                    #endif
+
+                    if (!telnetSTA) {
+                        DEBUG_MSG_P(PSTR("[TELNET] Rejecting - Only local connections\n"));
+                        _telnetDisconnect(i);
+                        return;
+                    }
+                }
+
+                _telnetNotifyConnected(i);
+
+                break;
+            }
+        }
+
+        //no free/disconnected spot so reject
+        if (i == TELNET_MAX_CLIENTS) {
+            DEBUG_MSG_P(PSTR("[TELNET] Rejecting - Too many connections\n"));
+        }
+    }
+
+    // Read data from clients
+    for (int i = 0; i < TELNET_MAX_CLIENTS; i++) {
+        while (_telnetClients[i] && _telnetClients[i]->available()) {
+            char data[512];
+            size_t len = _telnetClients[i]->available();
+            unsigned int r = _telnetClients[i]->readBytes(data, min(sizeof(data), len));
 
+            _telnetData(i, data, r);
+        }
+    }
+}
+#else
+void _telnetNewClient(AsyncClient *cl) {
```

Comment:
Having AsyncClient in the function definition screws up building on environments where ESPAsyncTCP modules are not enabled. That's why I added a void* and manually cast the parameter.

The error looks something like this:
```
code/espurna/espurna.ino:545:23: error: variable or field '_telnetNewClient' declared void
code/espurna/espurna.ino:545:23: error: 'AsyncClient' was not declared in this scope
code/espurna/espurna.ino:545:36: error: 'client' was not declared in this scope
```

Reply:
I changed it back in https://github.com/xoseperez/espurna/pull/1799/commits/da065b2b6e00bf9ecec0105327fe3037db6e490e

Reply:
What exactly is your config? When I noticed that, I tried turning off web,thingspeak,alexa and switching to pubsub for mqtt - it still compiled.
Both 1.5.0 and 2.2.2 platforms, pio 4 rc (python3, if that ends up being the reason), if it somehow breaks prototype generation.

Reply:
That's because the code in dev now has ESPAsyncTCP always included in prototypes, but I'm running your to-be-comitted OTA PR code which has this part removed: https://github.com/xoseperez/espurna/pull/1751/files#diff-52135e36de9de9ff791c1184a170b897

Reply:
Note that da065b2 does not build. 
How come ota_asynctcp.ino works then? It has AsyncClient in arguments.

And you are interested in inspecting the full source, C-z the running pio process and look at espurna/espurna.ino.cpp file

Reply:
...because ota_asynctcp includes espasync header. ok, so that looks like a bug with #1751 , not with telnet. I'd go back to AsyncClient in argument list and either:
Return ESPAsyncTCP.h to the prototypes regardless of the ifdef conditions
Add `class AsyncClient;` line

Reply:
Now that I think about it: I agree, makes sense to just keep it always in prototypes.h instead of making the code uglier - I changed it back. BTW it would make more sense if the generated espurna.ino.cpp would have macros around the forward declarations, but I don't think that's possible.